### ISSUE TITLE
no-restricted-imports: Unable to configure different messages for multiple named imports from same module 

### ISSUE THREAD

**Issue by hjylewis** [eslint/eslint#14220]
**Tell us about your environment**

* **ESLint Version:** v6.1.0
* **Node version:** v10.15.1
* **npm version:** v6.4.1

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**
babel-eslint
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
const path = require('path');

module.exports = {
	"root": true,
	"parser": "babel-eslint",
	"parserOptions": {
		"ecmaFeatures": {
			"jsx": true
		},
		"ecmaVersion": 6,
		"sourceType": "module"
	},

	"plugins": [
		"flowtype",
		"import",
		"jest",
		"jsx-a11y",
		"lodash",
		"prettier",
		"react",
		"react-hooks",
		"sprout",
		"no-jquery",
		"formatjs",
	],

	// We intentionally try to avoid extending other eslint configs for the following reasons:
	// 1. Explicit list of rules
	// 2. No unexpected rule changes from a package being upgraded
	// 3. Ability to include a comment about what each rule does so we don't have to go digging around
	"extends": [],

	"env": {
		"browser": true, // override with `node` for Node linting
		"commonjs": true,
		"es6": true
	},

	"globals": {
		"DEVELOPMENT": false, // replaced by build scripts
		"TESTING": false, // replaced by build scripts
		"module": false, // For Hot Module Reloading (module.hot.accepts)
		"Text": "off" // Must be explicitly used (i.e. window.Text) otherwise it will be confused with <Text/> from Racine
	},

	"rules": {
		// Import rules
		"no-restricted-imports": [
			"warn", 
			{ 
				"paths": [
					{
						"name": "@company/library",
						"importNames": ["toast"],
						"message": "Please use the default import from 'script/core/util/toast' instead."
					},
					{
						"name": "@company/library",
						"importNames": ["Numeral"],
						"message": "Please use `import {Numeral} from 'core/react';` instead."
					}
				]
			}
		]
	}
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
import {toast, Numeral} from '@company/library';
```

```bash
./node_modules/.bin/eslint file.js
```

**What did you expect to happen?**
Both the `toast` and `Numeral` imports should have warnings

**What actually happened? Please copy-paste the actual, raw output from ESLint.**
Only the `Numeral import has a warning.
```bash
   52:1   warning  '@company/library' import is restricted from being used. Please use `import {Numeral} from 'core/react';` instead  no-restricted-imports
```

**Steps to reproduce this issue:**

1. Configure `no-restricted-imports` to error on two named imports from the same module
1. Run eslint on file that imports both named imports

**Are you willing to submit a pull request to fix this bug?**
Potentially

**Comment by nzakas** [comment]
As of right now, this is behaving as the rule is intended to work. You can’t supply different error messages on a per-import basis, only on a per-module basis. Your second configuration is overwriting the first because you are using the same module name.

The open questions we have here are:

1. Does this overwriting make sense? It is how rule configs generally work but should it throw an error? Or should it combine the configs?
2. Do we want to support separate messages per import? The rule was never intended to do that and is already pretty complex. Is it worth additional complexity to add such a change?

**Comment by hjylewis** [comment]
In our use-case, it would be nice to have separate messages per import. We'd like to deprecate different imports for different reasons.

Right now to get unblocked, we've just combined the message into one long message that addresses both imports. That obviously won't scale much more beyond two imports though.

**Comment by nzakas** [comment]
Understood. As I said, the rule was never intended to do that, and I’m not sure if there’s an appetite on the team to make such a change.

@eslint/eslint-tsc I could use some feedback on the two points in my previous comment.

**Comment by btmills** [comment]
> 1. Does this overwriting make sense? It is how rule configs generally work but should it throw an error? Or should it combine the configs?

Normally when we talk about rule configs overwriting, that's the case of a merged config with `no-restricted-imports` configured multiple times in e.g. a plugin, a config, and possibly another cascaded config, where the most specific config takes precedence. Without knowing how the rule's implemented, I might've assumed like hjylewis did that as long as they're part of the same rule config entry, they wouldn't overwrite. In database terms, my mental model would have the unique constraint as (module, specifier), so (`@company/library`, `toast`) and (`@company/library`, `Numeral`) would be distinct.

> 2. Do we want to support separate messages per import? The rule was never intended to do that and is already pretty complex. Is it worth additional complexity to add such a change?

If we could get away with keying by (module, specifier) with minimal additional complexity, I'd be positively inclined. That might get too hairy, particularly if we open up the possibility of (`@company/library`, _) and (`@company/library`, `toast`) where we have to figure out the latter is more specific. Even worse would be supporting exceptions, so forbid all from `@company/library` _except_ `toast`, which I don't think we want to get into.

This doesn't answer the question directly, but I'd also be open to donating the rule to `eslint-plugin-import` if they're willing to take it.

**Comment by mdjermanovic** [comment]
> 1. Does this overwriting make sense? It is how rule configs generally work but should it throw an error? Or should it combine the configs?

This overwriting doesn't make sense. Given that the rule isn't intended to support multiple entries with the same module name, it should report a configuration error. A problem is that JSON Schema doesn't support that kind of validation (unique property value in an array of objects; on top of that, the array can be a mix of strings and objects).

> 2. Do we want to support separate messages per import? The rule was never intended to do that and is already pretty complex. Is it worth additional complexity to add such a change?

`importNames` option accepts arrays, so it looks like the design decision was to not support separate messages. It certainly does make sense to support this feature, but since there's a relatively simple alternative with the `no-restricted-syntax` rule, I'm more in favor of not adding complexity to `no-restricted-imports`.

```js
/* eslint no-restricted-syntax: ["error",
  {
    selector: "ImportDeclaration[source.value='@company/library'] > ImportSpecifier[imported.name='toast']",
    message: "Use toast from..."
  },
  {
    selector: "ImportDeclaration[source.value='@company/library'] > ImportSpecifier[imported.name='Numeral']",
    message: "Use Numeral from..."
  },
  {
    selector: "ImportDeclaration[source.value='@company/library'] > ImportNamespaceSpecifier",
    message: "Use named imports, because some parts of this library are restricted."
  }
]*/

import {
  toast, // error: Use toast from...
  Numeral, // error: Use Numeral from...
  foo // ok
} from '@company/library';

import * as CompanyLibrary from '@company/library'; // error: Use named imports...
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLFxuICB7XG4gICAgc2VsZWN0b3I6IFwiSW1wb3J0RGVjbGFyYXRpb25bc291cmNlLnZhbHVlPSdAY29tcGFueS9saWJyYXJ5J10gPiBJbXBvcnRTcGVjaWZpZXJbaW1wb3J0ZWQubmFtZT0ndG9hc3QnXVwiLFxuICAgIG1lc3NhZ2U6IFwiVXNlIHRvYXN0IGZyb20uLi5cIlxuICB9LFxuICB7XG4gICAgc2VsZWN0b3I6IFwiSW1wb3J0RGVjbGFyYXRpb25bc291cmNlLnZhbHVlPSdAY29tcGFueS9saWJyYXJ5J10gPiBJbXBvcnRTcGVjaWZpZXJbaW1wb3J0ZWQubmFtZT0nTnVtZXJhbCddXCIsXG4gICAgbWVzc2FnZTogXCJVc2UgTnVtZXJhbCBmcm9tLi4uXCJcbiAgfSxcbiAge1xuICAgIHNlbGVjdG9yOiBcIkltcG9ydERlY2xhcmF0aW9uW3NvdXJjZS52YWx1ZT0nQGNvbXBhbnkvbGlicmFyeSddID4gSW1wb3J0TmFtZXNwYWNlU3BlY2lmaWVyXCIsXG4gICAgbWVzc2FnZTogXCJVc2UgbmFtZWQgaW1wb3J0cywgYmVjYXVzZSBzb21lIHBhcnRzIG9mIHRoaXMgbGlicmFyeSBhcmUgcmVzdHJpY3RlZC5cIlxuICB9XG5dKi9cblxuaW1wb3J0IHtcbiAgdG9hc3QsIC8vIGVycm9yOiBVc2UgdG9hc3QgZnJvbS4uLlxuICBOdW1lcmFsLCAvLyBlcnJvcjogVXNlIE51bWVyYWwgZnJvbS4uLlxuICBmb28gLy8gb2tcbn0gZnJvbSAnQGNvbXBhbnkvbGlicmFyeSc7XG5cbmltcG9ydCAqIGFzIENvbXBhbnlMaWJyYXJ5IGZyb20gJ0Bjb21wYW55L2xpYnJhcnknOyAvLyBlcnJvcjogVXNlIG5hbWVkIGltcG9ydHMuLi4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)
